Support unconfirmed resource for nodes to improve scheduling fairness

TaoYang526 opened the pull request:
Currently testing results of node fairness seem not fair enough, as following graph shows.
![image](https://user-images.githubusercontent.com/15521649/68202365-8e8ea580-fffe-11e9-83b9-4e18902aeb05.png)
The root cause is that node sorting is not based on the latest node info, after an allocation proposal 
 is generated, available resource of related node will be deducted in ClusterInfo#processAllocationProposalEvent (handle AllocationProposalBundleEvent), in this gap, the scheduler maybe generating many allocation proposals based on outdated available resource of nodes.
I propose to add unconfirmed resource for node after allocation proposal is generated, then remove unconfirmed resource when allocation proposal is applied or rejected. So that we can sorting nodes  based on the latest available resource (=availableResource - unconfirmedResource).
After that, we can get a reasonable testing results of node fairness.
![image](https://user-images.githubusercontent.com/15521649/68203660-3dcc7c00-0001-11ea-9904-2a8911af4e9a.png)

yangwwei commented:
Hi @TaoYang526 

I agree. This is a nice catch.
The overall solution looks good, and also thanks for testing this out. Please see my detail comments in the commit.

yangwwei commented on pkg/cache/node_info.go:
```diff
@@ -29,6 +29,7 @@ type NodeInfo struct {
     TotalResource     *resources.Resource
     allocatedResource *resources.Resource
     availableResource *resources.Resource
+    unconfirmedResource *resources.Resource
```
can we rename this to `pendingAllocateResource`?

TaoYang526 commented on pkg/cache/node_info.go:
```diff
@@ -29,6 +29,7 @@ type NodeInfo struct {
     TotalResource     *resources.Resource
     allocatedResource *resources.Resource
     availableResource *resources.Resource
+    unconfirmedResource *resources.Resource
```
Yes, it make sense to me.

TaoYang526 commented on pkg/scheduler/allocator.go:
```diff
@@ -72,7 +72,10 @@ func (m *Scheduler) singleStepSchedule(nAlloc int, preemptionParam *preemptionPa
                 if err == nil {
-                    m.eventHandlers.CacheEventHandler.HandleEvent(newSingleAllocationProposal(alloc))
+                    // update unconfirmed resource of related node for new allocation proposal
+                    alloc.Node.NodeInfo.AddUnconfirmedResource(alloc.SchedulingAsk.AllocatedResource)
```
I think we should update _pendingAllocateResource_ of related node immediately after generating an allocation proposal, in that way we can guarantee an exact fairness of nodes, plan B is not to affect cache layer but to keep _pendingAllocateResources_ (structure like map[NodeId]*resources.Resource) inside scheduler layer such as allocator.go, we can join _pendingAllocateResources_ when creating new scheduling nodes for every scheduling process. A better way is to correlate scheduling node with node info, so that we can not create scheduling nodes for every scheduling process and just directly add _pendingAllocateResource_ into scheduling node. Thoughts?

yangwwei replied:
The reason to have a cache layer is to support multi-routine scheduling in the future, let's stay with that.  For the issue you mentioned, it probably won't be too bad. Have you tried this way? We can run the same test again to see if it is OK.

TaoYang526 replied:
Currently available resource of node is already deducted in ClusterInfo#processAllocationProposalEvent (further calling stack is PartitionInfo#addNewAllocation --> PartitionInfo#addNewAllocationInternal --> NodeInfo#addAllocation), so I think the first test results can prove this issue, am I right?

yangwwei replied:
Hi @TaoYang526 
Got what you are saying, I just want to make sure we are using the best solution to fix this. This is pretty critical.
So can we keep this in `SchedulerNode`, I noticed we already keep tracking of `AllocatingResource` in scheduler node, can we sort nodes with consideration of this part? We can change how to calculate the `CachedAvailableResource` in scheduler node, currently, it is 

```
info.GetAvailableResource()
```

it fetches the latest available resource from the cache layer. Instead, we can initiate an `AvailableResource` in the scheduler node with the initial value `info.GetAvailableResource()`, and then when fetching the latest value during sorting nodes, we do

```
AvailableResource = AvailableResource - AllocatingResource
```

would this work?
This way, we are not holding read lock from node_info (the cache layer), and in each scheduling interval, we maintain and update the nodes info in scheduler nodes.

Please let me know your thought.
Thanks

TaoYang526 replied:
Hi, @yangwwei 
Currently _AllocatingResource_  is only valuable for batch allocations which haven't been supported in once scheduling process, if we can correlate scheduling node with node info, I think it's better enough to take _AllocatingResource_ as _ pendingAllocateResource_, moreover we don't have to create new scheduling nodes for every scheduling process. Thoughts?

TaoYang526 commented:
Force pushed this PR to rebased from master.
@yangwwei  I have updated this PR as we discussed before, please help review and give some suggestions. Thanks.

yangwwei commented:
Thanks @TaoYang526 , overall it looks good. I just have some small things but I think we can track those later. Let's merge this in first. Thanks so much!

yangwwei commented:
Hi @TaoYang526 

Overall it looks good. I have one comment, 

The `CachedAvailableResource` in `SchedulingNode`, currently, it is updated in `schedulingNode#CheckAndAllocateResource`, this is called when we try to place every allocation and it requires to access the cache layer nodeInfo object. I would suggest to let scheduler manage the updates of the available resource by itself, that means when scheduler handles `SchedulerAllocationUpdatesEvent`, we probably calculate both `allocating` resource and `available` resource.  Does that make sense?

TaoYang526 commented:
@yangwwei Thanks for the review and comments, make sense to me.
I have updated this PR with 2 updates:
(1) Add cachedAvailableResource to SchedulingNode and maintain it internally.
(2) Lock in SchedulingNode#CheckAndAllocateResource is invalid since "defer" is lost before _m.lock.Unlock()_ , added it now.

yangwwei commented:
Thanks for the updates, it looks good to me! Merging this to master now.
